frappe/events_streaming/doctype/event_producer/event_producer.py

```diff
@@ -0,0 +1,359 @@
+			self.retry -= 1
+		if res.status_code == 200:
+				self.run_method(method)
+		if res.status_code != 200 and self.retry == 0:
+			frappe.throw(_('Failed to connect to the Event Producer site. Retry after some time.'))
+
+def get_current_node():
+	current_node = frappe.utils.get_url()
+	parts = current_node.split(':')
+	if not len(parts) > 2:
+		port = frappe.conf.http_port or frappe.conf.webserver_port
+		current_node += ':' + str(port)
+	return current_node
+
+def get_producer_site(producer_url):
+	producer_doc = frappe.get_doc('Event Producer', producer_url)
+	producer_site = FrappeClient(
+		url=producer_url,
+		api_key=producer_doc.api_key,
+		api_secret=producer_doc.get_password('api_secret'),
+		frappe_authorization_source='Event Consumer'
+	)
+	return producer_site
+
+def get_approval_status(config, ref_doctype):
+	for entry in config:
+		if entry.get('ref_doctype') == ref_doctype:
+			return entry.get('status')
+	return 'Pending'
+
+@frappe.whitelist()
+def pull_producer_data():
+	response = requests.get(get_current_node())
+	if response.status_code == 200:
+		'''Fetch data from producer node.'''
+		for event_producer in frappe.get_all('Event Producer'):
+			pull_from_node(event_producer.name)
+		return 'success'
+
+@frappe.whitelist()
+def pull_from_node(event_producer):
+	event_producer = frappe.get_doc('Event Producer', event_producer)
+	producer_site = get_producer_site(event_producer.producer_url)
+	last_update = event_producer.last_update
+
+	(doctypes, mapping_config, naming_config) = get_config(event_producer.event_configuration)
+
+	updates = get_updates(producer_site, last_update, doctypes)
+
+	for update in updates:
+		update.use_same_name = naming_config.get(update.ref_doctype)
+		mapping = mapping_config.get(update.ref_doctype)
+		if mapping:
+			update.mapping = mapping
+			update = get_mapped_update(update)
+		if not update.update_type == 'Delete':
+			update.data = json.loads(update.data)
+
+		sync(update, producer_site, event_producer)
+
+def get_config(event_config):
+	doctypes, mapping_config, naming_config = [], {}, {}
+
+	for entry in event_config:
+		if entry.status == 'Approved':
+			if entry.has_mapping:
+				(mapped_doctype, mapping) = frappe.db.get_value('Document Type Mapping', entry.mapping, ['remote_doctype', 'name'])
+				mapping_config[mapped_doctype] = mapping
+				naming_config[mapped_doctype] = entry.use_same_name
+				doctypes.append(mapped_doctype)
+			else:
+				naming_config[entry.ref_doctype] = entry.use_same_name
+				doctypes.append(entry.ref_doctype)
+	return (doctypes, mapping_config, naming_config)
+
+def sync(update, producer_site, event_producer, in_retry=False):
+	try:
+		if update.update_type == 'Create':
+			set_insert(update, producer_site, event_producer.name)
+		if update.update_type == 'Update':
+			set_update(update, producer_site)
+		if update.update_type == 'Delete':
+			set_delete(update)
+		if in_retry:
+			return 'Synced'
+		log_event_sync(update, event_producer.name, 'Synced')
```

Comment:
log clearing
- [ ] Create a single DocType "Log Settings"
- [ ] By default all logs get cleared in 30 days
- [ ] Create a child table for exception (DocType, Log Days)
- [ ] Refactor all code related to doctypes ending with "Log"

Reply:
Log Settings: https://github.com/frappe/frappe/pull/8788